### PR TITLE
Made RedLockMutexVessel usable with Redis cluster.

### PR DESCRIPTION
**Pull request by wingunder**
Up to now, a Redis reference had to be passed to RedLockMutexVessel at
construction time. This forced the user to pass the exact redis
instance that contains the key name, when using a cluster. If this was
not done, the cluster might tell the client that the key moved to
another node. As the Redis class does not know how to handle this, it
will result in an exception.

This commit turns RedLockMutexVessel into a template, that enables
the constructor to accept either a Redis or RedisCluster. This makes
the use of RedLockMutexVessel with or without clustering completely
transparent.

### CODE REVIEW

File: test/src/sw/redis++/redlock_test.hpp [hunk 1]
```diff
@@ -165,26 +160,26 @@ void RedLockTest<Redis>::run() {
...
+//        }
+//        while (mutex_list.size() != 0) {
+//            mutex_list.front()->unlock(random_string);
+//            delete(mutex_list.front());
+//            mutex_list.pop();
+//        }
+//    }
+//    end = std::chrono::system_clock::now();
+//    diff = end-start;
+//    std::cout << "Time to lock and unlock " << n << " simultaneous locks with RedMutex: " << diff.count() << " s" << std::endl;;
```
Comment (wingunder) [inline test/src/sw/redis++/redlock_test.hpp]: I didn't want to change RedMutex, so I just commented this test, until we are sure what we want to do here.

**Comment by sewenew** [comment]
Hi @wingunder 

First of all, sorry for the late reply...

> This forced the user to pass the exact redis instance that contains the key name, when using a cluster.

This problem can be solved with [hash-tag](https://redis.io/topics/cluster-spec#keys-hash-tags). If you have a `RedisCluster` instance, you can use the `RedisCluster::redis(const StringView &hash_tag)` method to get the exact Redis instance holding the given key. Then you can pass this `Redis` instance to `RedMutex`.

```
auto cluster = RedisCluster("tcp://127.0.0.1:7000");
std::string resource = "resource-name";
// Ge the Redis instance holding the resource.
auto redis = cluster.redis(resource);
RedMutex mtx(redis, resource);
RedLock<RedMutex> locker(mtx, std::defer_lock);
locker.try_lock(std::chrono::milliseconds(100));
```

In fact, I'm not sure if RedLock algorithm works with Redis Cluster. It seems [the official doc](https://redis.io/topics/distlock) doesn't mention any thing about Redis Cluster. I'm not quite sure if there's any side-effect when the [slot-node mapping changes](https://redis.io/commands/cluster-setslot#redis-cluster-live-resharding-explained). I also checked the implementation of [redlock-rb](https://github.com/antirez/redlock-rb/blob/master/redlock.rb). Redlock-rb doesn't support Redis Cluster either.

If the solution I mentioned above can solve your problem, I suggest pending this PR, and do more research on whether Redlock algorithm works with Redis Cluster.

Thanks again for your contribution to redis-plus-plus :)

Regards

**Comment by wingunder** [comment]
Hi @sewenew,
Thanks for your response and no problem regarding the delay.

I have a system that's using some pure Redis instances. We're now moving the stuff to a cluster, but it is required to keep the code backwards compatible with the our 'old' instance implementation. We also need to synchronize between pure Redis instances and a cluster. This is where redis-plus-plus really shines, as up to now I could just template everything. I then just decide which class (`sw::redis::Redis` or `sw::redis::RedisCluster`) to use at start-up, and everything just works, with no code changes or tricks. This is really great and a huge advantage of using `redis-plus-plus`! :)

The only places where it does not work seamlessly, are at the places where the `auto redis = cluster.redis(key);` trick is required (ping, info, parts of redlock, etc.). At all these places, I need to extract the relevant part and make 2 template instantiations, one for `sw::redis::Redis` and one for `sw::redis::RedisCluster`. It would really be nice if this could be done inside the `redis-plus-plus` library.

I was actually using the solution that you suggested, but I finally found it much easier to just change the redlock part of `redis-plus-plus` in the way I did in this PR.

Regarding the use of redlock with Redis cluster: It's working for me and the way I understand it, is that the cluster only uses the master node of a key to write to (not the key's slaves). After the key was written to the master node, the key gets replicated to its slave nodes. So, if one tries to write to a key, one always get redirected to a singe node (the master node for that key). This makes using redlock on a cluster pretty fast, as there's only one node that needs to get (red)locked. Reading is most probably different, but as my current use-case only requires lock-to-write, I haven't looked at lock-to-read.

I will completely understand if you don't want to integrate this patch, as I probably have a special use-case. Just let me know and then I'll close it.

**Comment by sewenew** [comment]
Hi @wingunder 

I have to admit the hash-tag trick is not very convenient. However, so far I don't have a good idea to avoid it.

About the Redlock algorithm and Redis Cluster, in your case, I think it doesn't work. The following is the reason (correct me if I'm wrong): 

Your master node has slave nodes, so if the master node fails, one of its slave will be promoted to be the new master by Redis Cluster. Say, client A has locked the Redlock, and wrote a key-value pair to the master node. However, before the key-value has been replicated to its slave, the master fails. Then one of its slave will be promoted to be the new master, and the new master doesn't  have the key-value pair. So client B can lock the Redlock with the new master. In this case, there will be two clients holding the lock. Also check the [doc](https://redis.io/topics/distlock#why-failover-based-implementations-are-not-enough) for detail.

You can remove the slaves to avoid this problem. However, as I mentioned above, I still not quite sure if the change of slot-node mapping, e.g. the resource key moved to another master node, has any side-effect to the Redlock algorithm.

I suggest you using a single Redis instance without slave (with single point failure problem), or multiple independent Redis instances without slaves (the [distributed and safer version](https://redis.io/topics/distlock#the-redlock-algorithm)) to implement the Redlock algorithm.

Regards

**Comment by wingunder** [comment]
Hi @sewenew 

> Your master node has slave nodes, so if the master node fails, one of its slave will be promoted to be the new master by Redis Cluster. Say, client A has locked the Redlock, and wrote a key-value pair to the master node. However, before the key-value has been replicated to its slave, the master fails. Then one of its slave will be promoted to be the new master, and the new master doesn't have the key-value pair. So client B can lock the Redlock with the new master. In this case, there will be two clients holding the lock. Also check the [doc](https://redis.io/topics/distlock#why-failover-based-implementations-are-not-enough) for detail.

Thanks. Yes, for sure the redlock that I suggested won't be a 100% safe during a master demotion.

The implementation I suggested is good enough for my use-case, as the only time it goes wrong is during fail-over. This is a minimal risk that is not critical for my use-case. It does require special care during manual cluster movements, but this is pretty easy to do from the application side. So the remaining risks are redis crashes (that I have not seen up to now), HW and/or network failures.

> I suggest you using a single Redis instance without slave (with single point failure problem), or multiple independent Redis instances without slaves (the [distributed and safer version](https://redis.io/topics/distlock#the-redlock-algorithm)) to implement the Redlock algorithm.

Thanks for the suggestion. This should give a 100% safe lock. To do this, we're back to the point where RedisCluster should know about all its masters _and_ slaves (probably #20). Once this is done, your suggestion could be implemented inside the Red* classes, and they maybe could be changed into templates.

At the moment I'll stick to what I have for my use-case.
Is it's ok with you if I now close this PR?

Regards

**Comment by sewenew** [comment]
Hi @wingunder 

Yes, you can close this PR :)

Regards

**Comment by wingunder** [comment]
Closing.

**Pull request closed without merging**